fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -179,69 +179,126 @@ private IndexingByIndex getIndexerByIndex() {
     }
 
     @Nonnull
-    private CompletableFuture<Void> handleIndexerReturnOrFallback(Throwable ex, Supplier<CompletableFuture<Void>> fallbackFunction) {
+    private CompletableFuture<Void> indexingLauncher(Supplier<CompletableFuture<Void>> indexingFunc) {
+        return indexingLauncher(indexingFunc, 0);
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> indexingLauncher(Supplier<CompletableFuture<Void>> indexingFunc, int attemptCount) {
+        return indexingLauncher(indexingFunc, attemptCount, null);
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> indexingLauncher(Supplier<CompletableFuture<Void>> indexingFunc, int attemptCount, @Nullable IndexingPolicy requestedPolicy) {
+        // The launcher calls the indexing function, letting the results be handled by the catcher.
+        // The catcher may, on some cases, call the launcher in its retry path. The attemptCount limits the recursion level as a safety net.
+        return AsyncUtil.composeHandle( indexingFunc.get(),
+                (ignore, ex) -> indexingCatcher(ex, indexingFunc, attemptCount + 1, requestedPolicy));
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> indexingCatcher(Throwable ex, Supplier<CompletableFuture<Void>> indexingFunc, int attemptCount, @Nullable IndexingPolicy requestedPolicy) {
+        // (skeleton function, a little long but broken to distinct cases)
         if (ex == null) {
+            // A happy index it is
             return AsyncUtil.DONE;
         }
-        IndexingBase.PartlyBuiltException partlyBuiltException = IndexingBase.getAPartlyBuildExceptionIfApplicable(ex);
+
+        if (attemptCount > 5) {
```

Comment:
I'm not sure it has any meaning for the caller. The max attempts, according to the algorithm, is 3 (attempt to continue previous by-index, attempt to rebuild requested by-index, then fallback to by-records). The `attemptCount` is only a recursion safety net in case of a bug. 
(Coming to think about it, it should probably trigger an error, not a warning).

Reply:
(Converted to error)

Reply:
Hm, I see. I think it's reasonable not to make it a configuration parameter, then, though I still think it should be a constant defined in the class (i.e., a `private static final` class member).

Reply:
Changed to: 
```if (attemptCount > INDEXING_ATTEMPTS_RECURSION_LIMIT)```